File: Modulo 2/Soluzione/dotnet/TaskManagement.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TaskManagementApp
{
    public class Task
    {
        public string Id { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public bool Completed { get; set; }
        public DateTime CreatedAt { get; set; }

        public Task()
        {
            Id = Guid.NewGuid().ToString();
            Completed = false;
            CreatedAt = DateTime.Now;
        }
    }

    public class TaskManager
    {
        private List<Task> tasks;

        public TaskManager()
        {
            tasks = new List<Task>();
        }

        public Task AddTask(string title, string description)
        {
            var newTask = new Task
            {
                Title = title,
                Description = description
            };
            
            tasks.Add(newTask);
            return newTask;
        }

        public bool RemoveTask(string id)
        {
            var task = tasks.FirstOrDefault(t => t.Id == id);
            if (task != null)
            {
                tasks.Remove(task);
                return true;
            }
            return false;
        }

        public bool CompleteTask(string id)
        {
            var task = tasks.FirstOrDefault(t => t.Id == id);
            if (task != null)
            {
                task.Completed = true;
                return true;
            }
            return false;
        }

        public List<Task> GetAllTasks()
        {
            return new List<Task>(tasks);
        }

        public List<Task> GetCompletedTasks()
        {
            return tasks.Where(t => t.Completed).ToList();
        }

        public List<Task> GetPendingTasks()
        {
            return tasks.Where(t => !t.Completed).ToList();
        }
    }
}

File: Modulo 2/Soluzione/typescript/TaskManagement.ts
export class Task {
    public id: string;
    public title: string;
    public description: string;
    public completed: boolean;
    public createdAt: Date;

    constructor() {
        this.id = this.generateId();
        this.completed = false;
        this.createdAt = new Date();
    }

    private generateId(): string {
        return Math.random().toString(36).substring(2, 11);
    }
}

export class TaskManager {
    private tasks: Task[];

    constructor() {
        this.tasks = [];
    }

    public addTask(title: string, description: string): Task {
        const newTask = new Task();
        newTask.title = title;
        newTask.description = description;
        this.tasks.push(newTask);
        return newTask;
    }

    public removeTask(id: string): boolean {
        const index = this.tasks.findIndex(t => t.id === id);
        if (index !== -1) {
            this.tasks.splice(index, 1);
            return true;
        }
        return false;
    }

    public completeTask(id: string): boolean {
        const task = this.tasks.find(t => t.id === id);
        if (task) {
            task.completed = true;
            return true;
        }
        return false;
    }

    public getAllTasks(): Task[] {
        return [...this.tasks];
    }

    public getCompletedTasks(): Task[] {
        return this.tasks.filter(t => t.completed);
    }

    public getPendingTasks(): Task[] {
        return this.tasks.filter(t => !t.completed);
    }
}
